File: modules/imgproc/misc/objc/test/MomentsTest.swift
//
//  MomentsTest.swift
//
//  Created by Giles Payne on 2020/02/10.
//

import XCTest
import OpenCV

class MomentsTest: XCTestCase {

    func testAll() {
        let data = Mat(rows: 3,cols: 3, type: CvType.CV_8UC1, scalar: Scalar(1))
        data.row(1).setTo(scalar: Scalar(5))
        let res = Imgproc.moments(array: data)
        XCTAssertEqual(res.m00, 21.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.m10, 21.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.m01, 21.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.m20, 35.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.m11, 21.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.m02, 27.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.m30, 63.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.m21, 35.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.m12, 27.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.m03, 39.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.mu20, 14.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.mu11, 0.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.mu02, 6.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.mu30, 0.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.mu21, 0.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.mu12, 0.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.mu03, 0.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.nu20, 0.031746031746031744, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.nu11, 0.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.nu02, 0.013605442176870746, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.nu30, 0.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.nu21, 0.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.nu12, 0.0, accuracy: OpenCVTestCase.EPS);
        XCTAssertEqual(res.nu03, 0.0, accuracy: OpenCVTestCase.EPS);
    }

}

File: modules/imgproc/misc/java/test/MomentsTest.java
package org.opencv.test.imgproc;

import org.opencv.test.OpenCVTestCase;
import org.opencv.core.Core;
import org.opencv.core.Mat;
import org.opencv.core.CvType;
import org.opencv.core.Scalar;
import org.opencv.imgproc.Imgproc;
import org.opencv.imgproc.Moments;

public class MomentsTest extends OpenCVTestCase {

    Mat data;

    @Override
    protected void setUp() throws Exception {
        super.setUp();

        data = new Mat(3,3, CvType.CV_8UC1, new Scalar(1));
        data.row(1).setTo(new Scalar(5));
    }

    public void testAll() {
        Moments res = Imgproc.moments(data);
        assertEquals(res.m00, 21.0, EPS);
        assertEquals(res.m10, 21.0, EPS);
        assertEquals(res.m01, 21.0, EPS);
        assertEquals(res.m20, 35.0, EPS);
        assertEquals(res.m11, 21.0, EPS);
        assertEquals(res.m02, 27.0, EPS);
        assertEquals(res.m30, 63.0, EPS);
        assertEquals(res.m21, 35.0, EPS);
        assertEquals(res.m12, 27.0, EPS);
        assertEquals(res.m03, 39.0, EPS);
        assertEquals(res.mu20, 14.0, EPS);
        assertEquals(res.mu11, 0.0, EPS);
        assertEquals(res.mu02, 6.0, EPS);
        assertEquals(res.mu30, 0.0, EPS);
        assertEquals(res.mu21, 0.0, EPS);
        assertEquals(res.mu12, 0.0, EPS);
        assertEquals(res.mu03, 0.0, EPS);
        assertEquals(res.nu20, 0.031746031746031744, EPS);
        assertEquals(res.nu11, 0.0, EPS);
        assertEquals(res.nu02, 0.013605442176870746, EPS);
        assertEquals(res.nu30, 0.0, EPS);
        assertEquals(res.nu21, 0.0, EPS);
        assertEquals(res.nu12, 0.0, EPS);
        assertEquals(res.nu03, 0.0, EPS);
    }

}
